SD_DSC.adoc

```diff
@@ -597,67 +597,22 @@ The evaluator checks the AGD documents to determine that any configuration that
+The algorithm tests require the developer to provide access to a test platform that provides the evaluator with tools that are typically not found on factory products.
 
-*Selected Long Messages Test, Bit-oriented Mode*
+The developer shall provide sufficient information to the evaluator to properly define the implementation of the algorithm. 
 
-The evaluators devise an input set consisting of 100 long messages ranging in size from rate+(rate+1) to rate+(100*(rate+1)), incrementing by rate+1. (For example, SHA-3-256 has a rate of 1088 bits. Therefore, 100 messages will be generated with lengths 2177, 3266, …, 109988 bits.) The message text shall be pseudo-randomly generated. The evaluators compute the message digest for each of the messages and ensure that the correct result is produced when the messages are provided to the TSF.
+All test results shall be compared to a known-good implementation to ensure the correct result is produced by the TSF. The evaluator shall determine the proper known-good implementation and provide a rationale for its use for testing. Internal or external implementations are acceptable with a sufficient rationale.
 
-*Selected Long Messages Test, Byte-oriented Mode*
+The tests are based on having a representative set of test vectors that exercise the algorithm. Within a single algorithm implementation there are many possible configuration parameters for how the algorithm is implemented. Some examples of these parameters can include:
 
-The evaluators devise an input set consisting of 100 messages ranging in size from (rate+(rate+8)) to (rate+100*(rate+8)), incrementing by rate+8. (For example, SHA-3-256 has a rate of 1088 bits. Therefore 100 messages will be generated of lengths 2184, 3280, 4376, …, 110688 bits.) The message text shall be pseudo-randomly generated. The evaluators compute the message digest for each of the messages and ensure that the correct result is produced when the messages are provided to the TSF.
+* input length
+* optional input variables 
+* output length
+* bit vs byte orientation
 
-*Pseudo-randomly Generated Messages (Monte Carlo) Test, Byte-oriented Mode*
+The evaluator shall generate a set of test vectors to exercise the algorithm according to the specified implementation details. Any parameters with a range shall test, at least, the minimum and maximum values specified for the parameters. The test vectors shall test each possible combination of the parameters. For example, if the algorithm has 2 optional input variables, a range of input lengths and 1 output length, then for each input length that is tested, each configuration of the optional input variables shall be tested (if the variables can be used together, that is another combination to be tested).
```

Comment:
This might be possible for some simple algorithms (e.g. AES with each padding mode, each key size, and maximum payload lengths). However, as more parameters become available, this becomes infeasible.

For example, for SP 800-56Br2 key transport (aka RSA OAEP), at least the following parameters exist:
- Role (initiator / responder)
- Key generation method (rsakpg1-basic, rsakpg1-prime-factor, rsakpg1-crt, rsakpg2-basic, rsakpg2-prime-factor, rsakpg2-crt)
- Modulo (2048, 3072, 4096, 6144, 8192 bits)
- Length of the key to derive (128-1024 bits)
- Hash algorithm (any of SHA-1, SHA-2, or SHA-3)
- Associated data pattern (if applicable)
- Key confirmation (a whole set of additional information)

SP 800-56Ar3 and SP 800-56Br2 define algorithms with even more parameters / combinations. The NIST ACVP server will not generate test vectors for all possible combinations. Instead, the server will generate 20 "test groups" that each test a random combination, and for each test group, 10 test vectors are included.

I suggest changing the sentence to "The test vectors shall test a reasonable number of combinations of the parameters to ensure coverage of the entire algorithm implementation."

Reply:
The following sentence should probably also be changed then, to "For example, if the algorithm has 2 optional input variables, a range of input lengths and 1 output length, then for each input length that is tested, each configuration of the optional input variables should be tested (if the variables can be used together, that is another combination to be tested)."

Reply:
@jvdsn Thank you. This is what I needed (since while I know how to do CAVP testing, I don't actually know what is going on underneath at the detailed level).

I have posted a new commit to try and bring this out.

It is possible we will need to tweak things based on the type of algorithm, which is annoying, but fine.